test/include/seqan3/test/tmp_directory.hpp

```diff
@@ -0,0 +1,198 @@
+// shipped with this file and also available at: https://github.com/seqan/seqan3/blob/master/LICENSE.md
+// -----------------------------------------------------------------------------------------------------
+
+#pragma once
+
+/*!\file
+ * \author Simon Gene Gottlieb <simon.gottlieb AT fu-berlin.de>
+ * \brief Internal test infrastructure.
+ *
+ * Define some helper classes and functions for the tests, that would be misplaced in the seqan3/include directory.
+ */
+
+#if defined(__APPLE__)
+#include <unistd.h>
+#elif defined(_WIN32)
+#include <cstring>
+#include <io.h>
+#else  // other unix systems
+#include <cstdlib>
+#endif
+
+#include <seqan3/std/filesystem>
+#include <iostream>
+
+#include <seqan3/core/platform.hpp>
+#include <seqan3/test/sandboxed_path.hpp>
+
+namespace seqan3::test
+{
+#if defined(_WIN32)
+namespace
+{
+/*!\brief Helper function that implements mkdtemp for windows.
+ *
+ * Caveat:
+ *
+ * Race condition between processes exists.
+ * There exists the possibility that between generating
+ * a unique file name and creating the directory another
+ * process also generates exactly the same filename.
+ */
+char * mkdtemp(char * template_name)
+{
+    if (_mktemp_s(template_name, strlen(template_name) + 1))
+        return nullptr;
+
+    if (std::filesystem::create_directories(template_name))
+        return template_name;
+
+    return nullptr;
+}
+}
+#endif
+
+//!\cond
+/*!\brief Creates and maintains a unique temporary directory.
+ *
+ * Creates a temporary unique directory. It automatically removes the temporary directory and all contained files and
+ * subdirectories on destruction. The class manages the life time of the associated directory. This means, when the
+ * instance is destructed the associated filesystem directory and all it's contents will be deleted automatically.
+ * Hence an instance of this class cannot be copied.
+ *
+ * ### Example
+ *
+ * \include test/snippet/test/tmp_directory.cpp
+ *
+ * ### Exceptions
+ *
+ * Might throw a std::filesystem::filesystem_error on failure to create a temporary file directory.
+ *
+ * ### Thread safety
+ *
+ * According to "https://www.gnu.org/software/libc/manual/html_node/Temporary-Files.html" the call to
+ * `mkdtemp` is thread safe, such that creating multiple parallel instances of this class will
+ * not induce a data race on the creation of temporary file path.
+ */
+class tmp_directory
+{
+public:
+    /* rule of six */
+    /*!\name Constructors, destructor and assignment
+     * \{
+     */
+    tmp_directory(tmp_directory const &) = delete; //!< Deleted.
+    tmp_directory(tmp_directory &&) = default; //!< Defaulted.
+    tmp_directory & operator=(tmp_directory const &) = delete; //!< Deleted.
+    tmp_directory & operator=(tmp_directory && other)
+    {
+        if (directory_path)
+        {
+            warn();
+            clean();
+        }
+        directory_path = std::move(other.directory_path);
+        return *this;
+    }
```

Comment:
```suggestion
    tmp_directory & operator=(tmp_directory && other)
    {
        if(this != &other)
            directory_path = std::exchange(other.directory_path, directory_path);
        return *this;
    }
```

I would just "swap" both objects, then the "other" object goes out of scope it will do that operation automatically.

See https://en.cppreference.com/w/cpp/utility/exchange#Notes